crates/shared/src/signature_validator/simulation.rs

```diff
@@ -0,0 +1,160 @@
+//! An `eth_call` simulation based ERC-1271 signature verification
+//! implementation. This allows orders with ERC-1271 signatures to be used that
+//! only get setup as a pre-hook (such as creating a Composable CoW order with a
+//! Safe in a pre-interaction).
+
+use {
+    super::{SignatureCheck, SignatureValidating, SignatureValidationError},
+    crate::code_simulation::{CodeSimulating, SimulationError},
+    anyhow::{Context, Result},
+    ethcontract::{common::abi::Token, tokens::Tokenize, Bytes},
+    ethrpc::extensions::StateOverride,
+    futures::future,
+    maplit::hashmap,
+    primitive_types::{H160, U256},
+    std::sync::Arc,
+    web3::types::CallRequest,
+};
+
+pub struct Validator {
+    simulator: Arc<dyn CodeSimulating>,
+    settlement: H160,
+    vault_relayer: H160,
+}
+
+impl Validator {
+    pub fn new(simulator: Arc<dyn CodeSimulating>, settlement: H160, vault_relayer: H160) -> Self {
+        Self {
+            simulator,
+            settlement,
+            vault_relayer,
+        }
+    }
+
+    async fn simulate(
+        &self,
+        check: &SignatureCheck,
+    ) -> Result<Verification, SignatureValidationError> {
+        // We simulate the signature verification from the Settlement contract's
+        // context. This allows us to check:
+        // 1. How the pre-interactions would behave as part of the settlement
+        // 2. Simulate the actual `isValidSignature` calls that would happen as part of
+        //    a settlement
+        let signatures = dummy_contract!(contracts::support::Signatures, self.settlement);
```

Comment:
I remember we made the `SettlementContract` inherit from `StorageAccessible` which should allow these types of simulations without requiring any fancy simulator/storage override if I'm not mistaken (using [`simulateDelegateCall`](https://github.com/cowprotocol/contracts/blob/main/src/contracts/mixins/StorageAccessible.sol#L60C14-L60C34)). 
Is there a particular reason we are not using this with a standard eth_call?

Would this also allow us to do TradeVerification on Gnosis Chain without a special node?

Reply:
I had completely forgotten about this! Yes, we can change the `Balances` and `Signatures` contracts to work this way.

It won't work with `TradeVerification` however, because that requires impersonating the trader and the solver, and not just the settlement contract.

Reply:
Will do this in a follow-up.